azurelinuxagent/common/protocol/util.py

```diff
@@ -317,34 +257,35 @@ def clear_protocol(self):
                 return
             logger.error("Failed to clear protocol endpoint: {0}", e)
 
-    def get_protocol(self, by_file=False):
+    def get_protocol(self):
         """
-        Detect protocol by endpoints, if by_file is True,
-        detect MetadataProtocol in priority.
+        Detect protocol by endpoint.
         :returns: protocol instance
         """
-
         if self._protocol is not None:
             return self._protocol
 
-        try:
-            self._protocol = self._get_protocol()
+        # If metadataserver certificates is present we clean certificates
+        # and remove MetadataServer firewall rule. It is possible
+        # Wire Protocol is being used due to a previous intermediate upgrade
+        # before 2.2.48 but metadata artifacts were not cleaned up.
+        if is_metadata_server_artifact_present():
```

Comment:
need to move this after _protocol_detect, if _protocol_detect fails we shouldn't delete MetadataServer certificates in the case where upgrade fails and we roll back to previous version.

Reply:
ok, could you also capture that scenario on a unit test? thanks